src/core/net/ip6.cpp

```diff
@@ -603,10 +618,254 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
+            payloadLeft -= maxPayloadFragment;
+            payloadFragment = maxPayloadFragment;
+        }
+        offset = fragmentCnt * (maxPayloadFragment >> 3);
+        fragmentHeader.SetOffset(offset);
+        VerifyOrExit((fragment = NewMessage(0)) != NULL, error = OT_ERROR_NO_BUFS);
+        SuccessOrExit(error = fragment->SetLength(aMessage.GetOffset() + sizeof(fragmentHeader) + payloadFragment));
+        header.SetPayloadLength(payloadFragment + sizeof(fragmentHeader));
+        assertValue = fragment->Write(0, sizeof(header), &header);
+        assert(assertValue == sizeof(header));
+        SuccessOrExit(error = fragment->SetOffset(aMessage.GetOffset()));
+        assertValue = fragment->Write(aMessage.GetOffset(), sizeof(fragmentHeader), &fragmentHeader);
+        assert(assertValue == sizeof(fragmentHeader));
+        VerifyOrExit(aMessage.CopyTo(aMessage.GetOffset() + (uint16_t)(offset << 3),
+                                     aMessage.GetOffset() + sizeof(fragmentHeader), payloadFragment,
+                                     *fragment) == (int)payloadFragment,
+                     error = OT_ERROR_NO_BUFS);
+        EnqueueDatagram(*fragment);
+
+        fragmentCnt++;
+        fragment = NULL;
+        otLogInfoIp6("Fragment %d with %d bytes sent ", fragmentCnt, payloadFragment);
+    }
+exit:
+    aMessage.Free();
+
+    if (error == OT_ERROR_NO_BUFS)
+    {
+        otLogWarnIp6("No buffer for Ip6 fragmentation");
+    }
+
+    if (error != OT_ERROR_NONE && fragment != NULL)
+    {
+        fragment->Free();
+    }
+
+    return error;
+}
+
+otError Ip6::HandleFragment(Message &    aMessage,
+                            Netif *      aNetif,
+                            Header &     aHeader,
+                            MessageInfo &aMessageInfo,
+                            bool         aFromNcpHost)
 {
     otError        error = OT_ERROR_NONE;
+    Header         header;
     FragmentHeader fragmentHeader;
+    uint16_t       offset          = 0;
+    uint16_t       payloadFragment = 0;
+    int            assertValue     = 0;
+    bool           isFragmented    = true;
+
+    VerifyOrExit(aMessage.Read(aMessage.GetOffset(), sizeof(fragmentHeader), &fragmentHeader) == sizeof(fragmentHeader),
+                 error = OT_ERROR_PARSE);
+
+    if (fragmentHeader.GetOffset() == 0 && fragmentHeader.IsMoreFlagSet() == false)
+    {
+        isFragmented = false;
+        error        = aMessage.MoveOffset(sizeof(fragmentHeader));
+    }
+    else if (mFragmentBuffer.mFragmentIdentification == 0 ||
+             (fragmentHeader.GetIdentification() == mFragmentBuffer.mFragmentIdentification &&
+              fragmentHeader.GetNextHeader() == mFragmentBuffer.mFragmentNextHeader))
+    {
+        offset          = (uint16_t)(fragmentHeader.GetOffset() << 3);
+        payloadFragment = aMessage.GetLength() - aMessage.GetOffset() - sizeof(fragmentHeader);
+        mFragmentBuffer.mReassemblyPayloadReceived += payloadFragment;
```

Comment:
Does this properly handle receiving a fragment multiple times (i.e. duplicates)?

Reply:
Very good point. No, that wouldn't work right now. Extending the simultaneous handling of different IPv6 datagrams will fix this.

Reply:
Commit https://github.com/openthread/openthread/pull/3948/commits/ccecde1d786f5f16a006357cf9934a739561bd07 should solve this problem.